### PR TITLE
Feat/add http destination connector

### PR DESCRIPTION
**Pull request by AnthonyR-AI2**
## Description

Adds an HTTP destination connector. This is a work in progress and comments and suggestions are welcome. 

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ x] New Connector (Destination or Source Connector)
- [ ] Breaking change (fix or feature that would impact existing functionality)
- [ ] Styling change
- [ ] Documentation update
- [ ] Refactoring
- [ ] Chore

## How Has This Been Tested?

Testing in progress...

## Checklist:

- [x ] Ensure a branch name is prefixed with `feature`, `bugfix`, `hotfix`, `release`, `style` or `chore` followed by `/` and branch name e.g `feature/add-salesforce-connector`
- [ x] Added unit tests for the changes made (if required)
- [x ] Have you made sure the commit messages meets the guidelines?
- [ ] Added relevant screenshots for the changes
- [x] Have you tested the changes on local/staging?
- [ ] Added the new connector in rollout.rb
- [ ] Have you updated the version number of the gem?
- [ ] Have you ensured that your changes for new connector are documented in the [docs repo](https://github.com/Multiwoven/docs) or relevant documentation files?
- [ ] Have you updated the run time dependency in multiwoven-integrations.gemspec if new gems are added
- [x] Have you made sure the code you have written follows the best practises to the best of your knowledge?

### CODE REVIEW

**Review by afthabvp — COMMENTED** [review]
Add http connector in integrations/lib/multiwoven/integrations.rb

File: .env.example [hunk 1]
```diff
@@ -1,63 +0,0 @@
-# Rails
-# ----------------------
```
Comment (afthabvp) [inline .env.example:2]: don't remove .env.example.
please revert this change

File: integrations/lib/multiwoven/integrations/destination/http/config/spec.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
+{
+  "documentation_url": "https://docs.multiwoven.com/integrations/destination/http",
+  "stream_type": "dynamic",
+  "connection_specification": {
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "title": "Http",
+    "type": "object",
+    "required": ["destination_url"],
```
Comment (afthabvp) [inline integrations/lib/multiwoven/integrations/destination/http/config/spec.json:8]: Is only the destination URL required for create an HTTP connection? Most HTTP clients use an API key or access token for authentication.
@subintp

File: integrations/lib/multiwoven/integrations/destination/http/client.rb [hunk 3]
```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
```
Comment (afthabvp) [inline integrations/lib/multiwoven/integrations/destination/http/client.rb:1]: discover method missing

File: integrations/lib/multiwoven/integrations/destination/http/config/spec.json (hunk 2 above)
Comment (AnthonyR-AI2) [reply integrations/lib/multiwoven/integrations/destination/http/config/spec.json:8]: This is a good question, I tossed around the idea of having an api key for authentication. I can certainly add it in if we require it.

File: integrations/lib/multiwoven/integrations/destination/http/client.rb (hunk 3 above)
Comment (AnthonyR-AI2) [reply integrations/lib/multiwoven/integrations/destination/http/client.rb:1]: I will add this.

File: integrations/lib/multiwoven/integrations/destination/http/client.rb [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+require "base64"
```
Comment (karthik-sivadas) [inline integrations/lib/multiwoven/integrations/destination/http/client.rb:3]: Can we add this in `integrations/lib/multiwoven/integrations.rb` this will help others also use it in integrations when need arises as it's a common library

File: integrations/lib/multiwoven/integrations/destination/http/config/spec.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    "properties": {
+      "destination_url": {
+        "type": "string",
+        "title": "Destination url",
+        "order": 0
+      },
+      "Username": {
```
Comment (karthik-sivadas) [inline integrations/lib/multiwoven/integrations/destination/http/config/spec.json:15]: It would be great if we add this inside `credentials` like the redshift as there is a high chance that we might need to accommodate various other authentication types like jwt tokens

File: integrations/lib/multiwoven/integrations/destination/http/client.rb [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  module Integrations
+    module Destination
+      module Http
+        include Multiwoven::Integrations::Core
+        class Client < DestinationConnector
+          MAX_CHUNK_SIZE = 10
+          def check_connection(connection_config)
+            connection_config = connection_config.with_indifferent_access
+            destination_url = connection_config[:destination_url]
+            request = Multiwoven::Integrations::Core::HttpClient.request(
```
Comment (subintp) [inline integrations/lib/multiwoven/integrations/destination/http/client.rb:14]: @AnthonyR-AI2  how are we confirming the credentials are valid by just sending an options request without any body ?

File: integrations/lib/multiwoven/integrations/destination/http/client.rb [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+            write_success = 0
+            write_failure = 0
+            records.each_slice(MAX_CHUNK_SIZE) do |chunk|
+              payload = create_payload(chunk)
+              response = Multiwoven::Integrations::Core::HttpClient.request(
+                url,
+                sync_config.stream.request_method,
+                payload: payload,
+                headers: auth_headers(auth_key)
```
Comment (subintp) [inline integrations/lib/multiwoven/integrations/destination/http/client.rb:57]: @AnthonyR-AI2  Auth headers needs to designed in a generic fashion. This is not extendable to different URL endpoints: What if the endpoint except` algolia-api-key` as header for the `auth_token`?

File: integrations/lib/multiwoven/integrations/destination/http/config/spec.json (hunk 5 above)
Comment (subintp) [inline integrations/lib/multiwoven/integrations/destination/http/config/spec.json:15]: @AnthonyR-AI2 Why are we expecting username and password using basic authentication supported by endpoint ? This is not generic implementation

**Review by subintp — CHANGES_REQUESTED** [review]

File: integrations/lib/multiwoven/integrations/destination/http/config/spec.json [hunk 8]
```diff
@@ -12,15 +12,10 @@
         "title": "Destination url",
         "order": 0
       },
-      "Username": {
-        "type": "string",
-        "title": "Authentication Username",
+      "headers": {
```
Comment (subintp) [inline integrations/lib/multiwoven/integrations/destination/http/config/spec.json:15]: How to FE knows these headers are an array of objects(key,value)?

**Review by subintp — CHANGES_REQUESTED** [review]

File: integrations/lib/multiwoven/integrations/core/constants.rb [hunk 9]
```diff
@@ -37,6 +37,7 @@ module Constants
       HTTP_POST = "POST"
       HTTP_PUT = "PUT"
       HTTP_DELETE = "DELETE"
+      HTTP_OPTIONS = "OPTIONS"
```
[the remark this answers is not in the document]
Comment (AnthonyR-AI2) [reply integrations/lib/multiwoven/integrations/core/constants.rb:40]: I added options as a way for us to check the connection of the HTTP endpoint. This should change given the discussions we've had about dynamic headers. I will make updates to all of this today .

File: integrations/lib/multiwoven/integrations/destination/http/client.rb (hunk 6 above)
Comment (AnthonyR-AI2) [reply integrations/lib/multiwoven/integrations/destination/http/client.rb:14]: Addressed this in my previous comment, this will change and will not be an options request.

File: integrations/lib/multiwoven/integrations/destination/http/config/spec.json (hunk 5 above)
Comment (AnthonyR-AI2) [reply integrations/lib/multiwoven/integrations/destination/http/config/spec.json]: We have removed this and now have dynamic headers.

File: integrations/lib/multiwoven/integrations/destination/http/config/spec.json (hunk 8 above)
Comment (AnthonyR-AI2) [reply integrations/lib/multiwoven/integrations/destination/http/config/spec.json:15]: This has been altered based on a conversation with FE on exact format.

File: integrations/lib/multiwoven/integrations/destination/http/client.rb (hunk 7 above)
Comment (AnthonyR-AI2) [reply integrations/lib/multiwoven/integrations/destination/http/client.rb]: These have been changed to generic headers.

File: integrations/lib/multiwoven/integrations/destination/http/client.rb (hunk 4 above)
Comment (AnthonyR-AI2) [reply integrations/lib/multiwoven/integrations/destination/http/client.rb:3]: removing this as it is not needed anymore

File: integrations/lib/multiwoven/integrations/destination/http/client.rb [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+            {
+              "records" => records.map do |record|
+                {
+                  "fields" => record
+                }
+              end
+            }
+          end
+
+          def convert_to_auth_key(username, password)
```
Comment (karthik-sivadas) [inline integrations/lib/multiwoven/integrations/destination/http/client.rb:88]: We can remove this if not needed

**Review by karthik-sivadas — APPROVED** [review]

**Comment by karthik-sivadas** [comment]
Closing this as this is a duplicate.
